northstar/src/runtime/state.rs

```diff
@@ -109,615 +89,464 @@ impl ProcessContext {
+        let manifest = if let Some(container) = self.containers.get(key) {
+            // Check if the container is not a resouce
+            if container.manifest.init.is_none() {
+                warn!("Container {} is a resource", key);
+                return Err(Error::UnknownApplication(key.clone()));
+            }
+
+            // Check if the container is already started
+            if container.process.is_some() {
+                warn!("Application {} is already running", key);
+                return Err(Error::ApplicationStarted(key.clone()));
+            }
 
-        // Check for all required resources
-        for mount in app.container.manifest.mounts.values() {
-            if let Mount::Resource { name, .. } = mount {
-                if !resources.contains(name) {
-                    return Err(Error::MissingResource(name.clone()));
+            container.manifest.clone()
+        } else if self.npks.contains_key(key) {
+            let container = self.mount(&key).await?;
+            let manifest = container.manifest.clone();
+            mounted.push(key.clone());
+            self.containers.insert(key.clone(), container);
+            manifest
+        } else {
+            warn!("Unknown application {}", key);
+            return Err(Error::UnknownApplication(key.clone()));
+        };
+
+        // Find to be mounted resources
+        for key in manifest.mounts.values().filter_map(|m| match m {
+            Mount::Resource {
+                name,
+                version,
+                repository,
+                ..
+            } => Some(ContainerKey::new(
+                name.clone(),
+                version.clone(),
+                repository.clone(),
+            )),
+            _ => None,
+        }) {
+            // Mount not yet mounted resources
+            if !self.containers.contains_key(&key) {
+                if self.npks.contains_key(&key) {
+                    // Obtain the key from the repo - needed for mounting
+                    match self.mount(&key).await {
+                        Ok(container) => {
+                            // Fine. Add the container to the list of mounted containers
+                            mounted.push(key.clone());
+                            self.containers.insert(key, container);
+                        }
+                        Err(e) => {
+                            warn!("Failed to mount {}", key);
+                            // Umount everything mounted for this particular start operation
+                            for k in mounted.drain(..) {
+                                self.umount(&k).await.ok();
+                            }
+                            return Err(e);
+                        }
+                    }
+                } else {
+                    for k in mounted.drain(..) {
+                        self.umount(&k).await.ok();
+                    }
+                    return Err(Error::UnknownResource(key));
```

Comment:
handling error could also be easier with a dedicated mount-required-resources function that can fail.
not so many loose ends to tie up

Reply:
The mounting code will be rewritten soon. See previous comment - so I didn't care much about it.
Agree that this is too noisy.

Reply:
Updated the patch with the following behaviour: Whenever a mount fails the container is kept mounted. In a worst case the rt would need to be abled to handle every present npk to be mounted so a umount is not needed upon start errors.